qt6.wrapQtAppsHook: add `qtwayland` to `propagatedBuildInputs`

Artturin opened the pull request:
Cont of https://github.com/NixOS/nixpkgs/pull/333012

- **qt6.wrapQtAppsHook: add `qtwayland` to `propagatedBuildInputs`**
- **qt6.wrapQtAppsNoGuiHook: init**
- **qt6.qtwayland: set meta.{platforms,badPlatforms}**
- **doc/qt: mention wrapQtAppsNoGuiHook**
- **qtbase-setup-hook: add wrapQtAppsNoGuiHook to error message**
- **deepin.dde-application-manager: use wrapQtAppsNoGuiHook to avoid propagating qtwayland**
- **pineapple-pictures: let wrapQtAppsHook propagate qtwayland**

wineee committed:
qt6.wrapQtAppsHook: add `qtwayland` to `propagatedBuildInputs`

qtwayland client side is the wayland platform plugin, provides a way to run Qt applications as Wayland clients
without this, Qt applications only work on x11/xwayland

Co-authored-by: Artturin <Artturin@artturin.com>

wineee committed:
qt6.wrapQtAppsNoGuiHook: init

Unlike wrapQtAppsHook, wrapQtAppsNoGuiHook does not propagate qtwayland,
to reduce closure size for CLI or other non-GUI applications.

Co-authored-by: eclairevoyant <contactmeongithubinstead@proton.me>

wineee committed:
qt6.qtwayland: set meta.{platforms,badPlatforms}

wineee committed:
doc/qt: Mention propagates and wrapQtAppsNoGuiHook

Co-authored-by: eclairevoyant <contactmeongithubinstead@proton.me>
Co-authored-by: Artturin <Artturin@artturin.com>

wineee committed:
qtbase-setup-hook: add wrapQtAppsNoGuiHook to error message

Co-authored-by: eclairevoyant <contactmeongithubinstead@proton.me>

wineee committed:
deepin.dde-application-manager: use wrapQtAppsNoGuiHook to avoid propagating qtwayland

Co-authored-by: eclairevoyant <contactmeongithubinstead@proton.me>

wineee committed:
pineapple-pictures: let wrapQtAppsHook propagate qtwayland

K900 commented on doc/languages-frameworks/qt.section.md:
```diff
@@ -25,12 +25,14 @@ stdenv.mkDerivation {
-This may become default in the future, see [NixOS/nixpkgs#269674](https://github.com/NixOS/nixpkgs/pull/269674).
+`wrapQtAppsHook` will propagate `qtwayland` to ensure the Wayland platform plugin is available, which is useful for graphical applications.
+
+`wrapQtAppsNoGuiHook` will not propagate any qt components to reduce closure size for purely command-line applications.
```
Qt

K900 replied:
Also, maybe "anything but qtbase"? qtbase is technically a Qt component

NickCao replied:
"additional qt components", I propose.

Artturin replied:
> Also, maybe "anything but qtbase"? qtbase is technically a Qt component

Well it does not propagate that currently?

K900 reviewed (approved):
LGTM minus the nit

emilazy commented on pkgs/development/libraries/qt-6/default.nix:
```diff
@@ -167,10 +167,19 @@ let
           {
             name = "wrap-qt6-apps-hook";
             propagatedBuildInputs = [ makeBinaryWrapper ];
+            depsTargetTargetPropagated = lib.optionals (lib.meta.availableOn stdenv.targetPlatform qtwayland) [ qtwayland.dev ];
```
Just checking: `.dev` is definitely correct here?

K900 replied:
I think qtbase is still necessary, so the plugins from it can be found.

Artturin replied:
> I think qtbase is still necessary, so the plugins from it can be found.

Both of these launch
```
$ env -i HOME="$HOME" DISPLAY="$DISPLAY" WAYLAND_DISPLAY="$WAYLAND_DISPLAY" XDG_RUNTIME_DIR="$XDG_RUNTIME_DIR" QT_QPA_PLATFORM=xcb PATH="$PATH" QT_PLUGIN_PATH=/nix/store/kgz360i4bwc19c35ag7m35vn7f0910k2-qtwayland-6.8.0/lib/qt-6/plugins ./result/bin/.ppic-wrapped
$ env -i HOME="$HOME" DISPLAY="$DISPLAY" WAYLAND_DISPLAY="$WAYLAND_DISPLAY" XDG_RUNTIME_DIR="$XDG_RUNTIME_DIR" QT_QPA_PLATFORM=wayland PATH="$PATH" QT_PLUGIN_PATH=/nix/store/kgz360i4bwc19c35ag7m35vn7f0910k2-qtwayland-6.8.0/lib/qt-6/plugins ./result/bin/.ppic-wrapped
```

K900 replied:
I guess it's fine to just have it pulled from the application's dependencies.

emilazy replied:
Propagated build inputs only propagate one level, right? I can conceive of some package that encapsulates its use of Qt so that the actual application doesn’t have an explicit `qtbase` dependency and would thus break.

Artturin replied:
It would be good if we only propagated the plugins dir

Artturin replied:
Well that isn't possible because the plugins depend on $out

emilazy replied:
It seems fine just to propagate `qtbase.out` and `qtwayland.out` for now, right? I doubt the closure size from the header files matters much?

Artturin replied:
I'm not concerned with closure size, instead I'm concerned with env pollution

Artturin replied:
https://github.com/NixOS/nixpkgs/pull/352466

emilazy commented:
Want me to do a treewide after this?

Artturin commented:
Needs https://github.com/NixOS/nixpkgs/pull/352466

SuperSandro2000 commented:
> Want me to do a treewide after this?

I think it is hard to figure out if the package just had qtwayland because of this or because it genuinely needs it to compile.

SuperSandro2000 commented on pkgs/development/libraries/qt-6/default.nix:
Shouldn't we fix up the 2nd change when switching to availableOn?

Artturin replied:
Fixup [qt6.qtwayland: set meta.{platforms,badPlatforms}](https://github.com/NixOS/nixpkgs/pull/352419/commits/525338a3cf6b3ccf3ed9a41c7319deb8d8f698c4) to [qt6.wrapQtAppsHook: add qtwayland to propagatedBuildInputs](https://github.com/NixOS/nixpkgs/pull/352419/commits/7eb37716f3e74664226012163cc2d55843fcfd75) ?

SuperSandro2000 commented on doc/languages-frameworks/qt.section.md:
```diff
@@ -25,12 +25,14 @@ stdenv.mkDerivation {
 The same goes for Qt 5 where libraries and tools are under `libsForQt5`.
 
-Any Qt package should include `wrapQtAppsHook` in `nativeBuildInputs`, or explicitly set `dontWrapQtApps` to bypass generating the wrappers.
+Any Qt package should include `wrapQtAppsHook` (or `wrapQtAppsNoGuiHook`) in `nativeBuildInputs`, or explicitly set `dontWrapQtApps` to bypass generating the wrappers.
```
```suggestion
Any Qt package should include `wrapQtAppsHook` or `wrapQtAppsNoGuiHook` in `nativeBuildInputs`, or explicitly set `dontWrapQtApps` to bypass generating the wrappers.
```

SuperSandro2000 commented on doc/languages-frameworks/qt.section.md:
```diff
@@ -25,12 +25,14 @@ stdenv.mkDerivation {
-Qt 6 graphical applications should also include `qtwayland` in `buildInputs` on Linux (but not on platforms e.g. Darwin, where `qtwayland` is not available), to ensure the Wayland platform plugin is available.
 
-This may become default in the future, see [NixOS/nixpkgs#269674](https://github.com/NixOS/nixpkgs/pull/269674).
+`wrapQtAppsHook` will propagate `qtwayland` to ensure the Wayland platform plugin is available, which is useful for graphical applications.
```
```suggestion
`wrapQtAppsHook` propagates `qtwayland` to ensure the Wayland platform plugin is available, which is required for graphical applications to run under wayland on none Qt based desktops.
```

emilazy commented:
> I think it is hard to figure out if the package just had qtwayland because of this or because it genuinely needs it to compile.

I suspect there are extremely few users of the Qt Wayland Compositor API in‐tree, maybe even zero. But `nixpkgs-review` and `git log` can answer these questions.

SuperSandro2000 commented on pkgs/development/libraries/qt-6/hooks/qtbase-setup-hook.sh:
```diff
@@ -75,9 +75,9 @@ else # Only set up Qt once.
+        # Check that wrapQtAppsHook/wrapQtAppsNoGuiHook is used, or it is explicitly disabled.
         if [[ -z "$__nix_wrapQtAppsHook" && -z "$dontWrapQtApps" ]]; then
-            echo >&2 "Error: wrapQtAppsHook is not used, and dontWrapQtApps is not set."
+            echo >&2 "Error: wrapQtAppsHook and wrapQtAppsNoGuiHook are not used, and dontWrapQtApps is not set."
```
```suggestion
            echo >&2 "Error: neither wrapQtAppsHook or wrapQtAppsNoGuiHook are used, and dontWrapQtApps is not set either."
```

SuperSandro2000 commented:
> I suspect there are extremely few users of the Qt Wayland Compositor API in‐tree, maybe even zero. But `nixpkgs-review` and `git log` can answer these questions.

Oh, they wouldn't be mixed into buildInputs and cause impurities? That would be perfect for a treewide cleanup.

emilazy commented:
@Artturin is separating out the plugins in the other PR, so if that’s merged and we use that separate output then there shouldn’t be any risk of the non‐plugin APIs leaking.

wineee commented:
> I suspect there are extremely few users of the Qt Wayland Compositor API in‐tree, maybe even zero. But `nixpkgs-review` and `git log` can answer these questions.

At least `deepin.dde-shell` uses the Qt Wayland Compositor API，which uses the wayland protocol to load plugins, allowing plugins to run in separate processes

But there is probably no other users，qtwayland compositor once had [a serious bug](https://github.com/NixOS/nixpkgs/pull/337913), but it did not receive any user feedback

Artturin committed:
qt6.wrapQtAppsHook: Only propagate plugins,qml of `qtwayland`

qt6.{wrapQtAppsHook,wrapQtAppsNoGuiHook}: propagate `qtbase` plugins

https://www.github.com/NixOS/nixpkgs/pull/352419#discussion_r1823132984

K900 commented on doc/languages-frameworks/qt.section.md:
```diff
@@ -25,12 +25,14 @@ stdenv.mkDerivation {
-Qt 6 graphical applications should also include `qtwayland` in `buildInputs` on Linux (but not on platforms e.g. Darwin, where `qtwayland` is not available), to ensure the Wayland platform plugin is available.
 
-This may become default in the future, see [NixOS/nixpkgs#269674](https://github.com/NixOS/nixpkgs/pull/269674).
+`wrapQtAppsHook` propagates `qtwayland` to ensure the Wayland platform plugin is available, which is required for graphical applications to run under wayland on non-Qt based desktops.
```
It doesn't do that anymore, does it?

K900 commented on pkgs/development/libraries/qt-6/default.nix:
```diff
@@ -185,9 +185,28 @@ let
             name = "wrap-qt6-apps-hook";
+            propagatedBuildInputs = [ makeBinaryWrapper ];
+              depsTargetTargetPropagated = [
+              (runCommand "${qtbase.__spliced.targetTarget.name or qtbase.name}-only-plugins" { } ''
```
Maybe extract this into a function?

K900 commented on pkgs/development/libraries/qt-6/default.nix:
```diff
@@ -185,9 +185,28 @@ let
+              depsTargetTargetPropagated = [
+              (runCommand "${qtbase.__spliced.targetTarget.name or qtbase.name}-only-plugins" { } ''
+                mkdir -p "$out/lib/qt-6/plugins"
+                ln -s "${qtbase.__spliced.targetTarget or qtbase}/lib/qt-6/plugins" "$out/lib/qt-6/plugins"
```
Also, I'm not sure it matters here, but generally we could also want to propagate $out/lib/qt-6/qml.

K900 commented on pkgs/development/libraries/qt-6/hooks/qtbase-setup-hook.sh:
```diff
@@ -75,9 +75,9 @@ else # Only set up Qt once.
+        # Check that wrapQtAppsHook/wrapQtAppsNoGuiHook is used, or it is explicitly disabled.
         if [[ -z "$__nix_wrapQtAppsHook" && -z "$dontWrapQtApps" ]]; then
-            echo >&2 "Error: wrapQtAppsHook is not used, and dontWrapQtApps is not set."
+            echo >&2 "Error: neither wrapQtAppsHook or wrapQtAppsNoGuiHook are used, and dontWrapQtApps is not set either."
```
nor

K900 commented on doc/languages-frameworks/qt.section.md:
```diff
@@ -25,12 +25,14 @@ stdenv.mkDerivation {
-Qt 6 graphical applications should also include `qtwayland` in `buildInputs` on Linux (but not on platforms e.g. Darwin, where `qtwayland` is not available), to ensure the Wayland platform plugin is available.
 
-This may become default in the future, see [NixOS/nixpkgs#269674](https://github.com/NixOS/nixpkgs/pull/269674).
+`wrapQtAppsHook` propagates plugins and QML from `qtbase` and `qtwayland` to ensure the Wayland platform plugin is available, which is required for graphical applications to run under wayland on non-Qt based desktops. On platforms that do not support `qtwayland` (e.g. Darwin), only plugins and QML `qtbase` will be propagated.
```
qtbase can't contain QML, as it's a dependency of qtdeclarative. Also, "QML components" is probably the more correct term. Also, Wayland, capitalized. Sorry if I'm nitpicking too much.

K900 commented on doc/languages-frameworks/qt.section.md:
```diff
@@ -25,12 +25,14 @@ stdenv.mkDerivation {
-This may become default in the future, see [NixOS/nixpkgs#269674](https://github.com/NixOS/nixpkgs/pull/269674).
+`wrapQtAppsHook` propagates plugins and QML from `qtbase` and `qtwayland` to ensure the Wayland platform plugin is available, which is required for graphical applications to run under wayland on non-Qt based desktops. On platforms that do not support `qtwayland` (e.g. Darwin), only plugins and QML `qtbase` will be propagated.
+
+`wrapQtAppsNoGuiHook` will not propagate any Qt components to reduce closure size for purely command-line applications.
```
Propagating qtbase in noGuiHook is probably fine as it's basically impossible to write a Qt app without it.

emilazy replied:
You could be using a cross‐platform library that abstracts over Qt entirely. (Admittedly less likely for a CLI app.)

K900 replied:
You'll still have qtbase in the closure.

K900 commented on pkgs/development/libraries/qt-6/default.nix:
```diff
@@ -185,10 +185,31 @@ let
+            onlyPlugins = drv: (runCommand "${drv.__spliced.targetTarget.name or drv.name}-only-plugins" { } ''
+                mkdir -p "$out/lib/qt-6"
+                ln -s "${drv.__spliced.targetTarget or drv}/lib/qt-6/plugins" "$out/lib/qt-6/plugins"
+                ln -s "${drv.__spliced.targetTarget or drv}/lib/qt-6/qml" "$out/lib/qt-6/plugins"
```
Wrong path here, can also use `qtbase.qtPluginPrefix`/`qtbase.qtQmlPrefix` to avoid duplicating.

K900 commented on doc/languages-frameworks/qt.section.md:
```diff
@@ -25,12 +25,14 @@ stdenv.mkDerivation {
-Qt 6 graphical applications should also include `qtwayland` in `buildInputs` on Linux (but not on platforms e.g. Darwin, where `qtwayland` is not available), to ensure the Wayland platform plugin is available.
 
-This may become default in the future, see [NixOS/nixpkgs#269674](https://github.com/NixOS/nixpkgs/pull/269674).
+`wrapQtAppsHook` propagates plugins and QML components from `qtwayland` to ensure the Wayland platform plugin is available, which is required for graphical applications to run under Wayland on non-Qt based desktops. On platforms that do not support `qtwayland` (e.g. Darwin), only plugins `qtbase` will be propagated.
```
plugins _from_ `qtbase`?

K900 commented on doc/languages-frameworks/qt.section.md:
```diff
@@ -25,12 +25,14 @@ stdenv.mkDerivation {
-This may become default in the future, see [NixOS/nixpkgs#269674](https://github.com/NixOS/nixpkgs/pull/269674).
+`wrapQtAppsHook` propagates plugins and QML components from `qtwayland` to ensure the Wayland platform plugin is available, which is required for graphical applications to run under Wayland on non-Qt based desktops. On platforms that do not support `qtwayland` (e.g. Darwin), only plugins `qtbase` will be propagated.
+
+`wrapQtAppsNoGuiHook` does not propagate `qtwayland` reduce closure size for purely command-line applications.
```
_to_ reduce closure size

K900 commented on pkgs/development/libraries/qt-6/default.nix:
```diff
@@ -37,6 +38,12 @@ let
+
+      onlyPluginsAndQml = drv: (runCommand "${drv.__spliced.targetTarget.name or drv.name}-only-plugins" { } ''
+          mkdir -p "$out/lib/qt-6"
+          ln -s "${drv.__spliced.targetTarget or drv}/${self.qtbase.qtPluginPrefix}" "$out/lib/${self.qtbase.qtPluginPrefix}"
```
No /lib

K900 commented on pkgs/development/libraries/qt-6/default.nix:
```diff
@@ -37,6 +38,12 @@ let
       ];
+
+      onlyPluginsAndQml = drv: (runCommand "${drv.__spliced.targetTarget.name or drv.name}-only-plugins" { } ''
+          mkdir -p "$out/lib/qt-6"
```
Maybe `mkdir -p $qtPluginPrefix` here to be sure?

K900 commented on pkgs/development/libraries/qt-6/hooks/qtbase-setup-hook.sh:
```diff
@@ -75,9 +75,9 @@ else # Only set up Qt once.
+        # Check that wrapQtAppsHook/wrapQtAppsNoGuiHook is used, or it is explicitly disabled.
         if [[ -z "$__nix_wrapQtAppsHook" && -z "$dontWrapQtApps" ]]; then
-            echo >&2 "Error: wrapQtAppsHook is not used, and dontWrapQtApps is not set."
+            echo >&2 "Error: neither wrapQtAppsHook nor wrapQtAppsNoGuiHook are used, and dontWrapQtApps is not set either."
```
At this point I wonder if this is better written as an imperative. "something something add wrapQtApps(NoGui)Hook or set dontWrapQtApps"?

K900 committed:
qt: editing pass on docs

vcunat commented:
I'll assume that you can finish this up within a day or two.

Artturin commented on pkgs/development/libraries/qt-6/default.nix:
```diff
@@ -37,6 +38,15 @@ let
+        inherit (self.qtbase) qtPluginPrefix qtQmlPrefix;
+      in (runCommand "${drv'.name}-only-plugins-qml" { } ''
+          mkdir -p $(dirname "$out/${qtPluginPrefix}")
+          test -d "${drv'}/${qtPluginPrefix}" && ln -s "${drv'}/${qtPluginPrefix}" "$out/${qtPluginPrefix}" || true
+          test -d "${drv'}/${qtQmlPrefix}" && ln -s "${drv'}/${qtQmlPrefix}" "$out/${qtQmlPrefix}" || true
```
fixed dangling symlink in qtbase
